consensus/core/src/leader_scoring.rs

```diff
@@ -270,12 +341,183 @@ impl ScoringSubdag {
+            .iter()
+            .enumerate()
+            .flat_map(|(subdag_index, subdag)| {
+                committed_leaders.push(subdag.leader);
+                if subdag_index == 0 {
+                    subdag.blocks.iter()
+                } else {
+                    let previous_subdag = &subdags[subdag_index - 1];
+                    let expected_next_subdag_index = previous_subdag.commit_ref.index + 1;
+                    assert_eq!(
+                        subdag.commit_ref.index, expected_next_subdag_index,
+                        "Non-contiguous commit index (expected: {}, found: {})",
+                        expected_next_subdag_index, subdag.commit_ref.index
+                    );
+                    subdag.blocks.iter()
+                }
+            })
+            .map(|block| (block.reference(), block.clone()))
+            .collect::<BTreeMap<_, _>>();
+
+        // Guaranteed to have a contiguous list of commit indices
+        let commit_range = CommitRange::new(
+            subdags.first().unwrap().commit_ref.index..=subdags.last().unwrap().commit_ref.index,
+        );
+
+        assert!(
+            !blocks.is_empty(),
+            "Attempted to create UnscoredSubdag with no blocks"
+        );
+
+        Self {
+            context,
+            commit_range,
+            committed_leaders,
+            blocks,
+        }
+    }
+
+    pub(crate) fn find_supported_leader_block(
+        &self,
+        leader_slot: Slot,
+        from: &VerifiedBlock,
+    ) -> Option<BlockRef> {
+        if from.round() < leader_slot.round {
+            return None;
+        }
+        for ancestor in from.ancestors() {
+            if Slot::from(*ancestor) == leader_slot {
+                return Some(*ancestor);
+            }
+            // Weak links may point to blocks with lower round numbers than strong links.
+            if ancestor.round <= leader_slot.round {
+                continue;
+            }
+            if let Some(ancestor) = self.get_block(ancestor) {
+                if let Some(support) = self.find_supported_leader_block(leader_slot, &ancestor) {
+                    return Some(support);
+                }
+            } else {
+                // TODO: Add unit test for this case once dagbuilder is ready.
+                tracing::trace!(
+                    "Potential vote's ancestor block not found in unscored committed subdags: {:?}",
+                    ancestor
+                );
+                return None;
+            }
+        }
+        None
+    }
+
+    pub(crate) fn is_vote(
+        &self,
+        potential_vote: &VerifiedBlock,
+        leader_block: &VerifiedBlock,
+    ) -> bool {
+        let reference = leader_block.reference();
+        let leader_slot = Slot::from(reference);
+        self.find_supported_leader_block(leader_slot, potential_vote) == Some(reference)
+    }
+
+    // TODO(arun): Remove
+    // pub(crate) fn is_certificate(
```

Comment:
nit: I guess since this is not being used at all anywhere we can even remote it as of this PR